### PR TITLE
iox-#14 replace ChunkInfo with ChunkHeader members

### PR DESCRIPTION
**Pull request by elBoberido**
Signed-off-by: Mathias Kraus <mathias.kraus@apex.ai>

## Pre-Review Checklist for the PR Author

1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Notes for Reviewer
This PR replaces `ChunkInfo` with members from `ChunkHeader` and moves some types to `iceoryx_posh_types.hpp`

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [x] All checkboxes in the PR checklist are checked or crossed out
1. [x] Merge

## References

- Partly closes #14

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp [hunk 1]
```diff
@@ -196,6 +196,18 @@ using ConfigFilePathString_t = cxx::string<1024>;
...
+namespace mepoo
+{
+using SequenceNumber_t = std::uint64_t;
+using BaseClock_t = std::chrono::steady_clock;
+
+// use signed integer for duration;
+// there is a bug in gcc 4.8 which leads to a wrong calcutated time
+// when sleep_until() is used with a timepoint in the past
+using DurationNs_t = std::chrono::duration<std::int64_t, std::nano>;
+using TimePointNs_t = std::chrono::time_point<BaseClock_t, DurationNs_t>;
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:208]: these could probably also fit into the plain `iox` namespace. It's a bit weird to have a `DurationNs_t` which is not based on `cxx::Duration` though

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl [hunk 2]
```diff
@@ -45,12 +45,11 @@ ChunkSender<ChunkSenderDataType>::tryAllocate(const uint32_t payloadSize, const
     const uint32_t neededChunkSize = getMembers()->m_memoryMgr->sizeWithChunkHeaderStruct(payloadSize);
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl:45]: with `sizeWithChunkHeaderStruct` we must be careful to also consider the optional CustomChunkHeader

**Review by budrus — DISMISSED** [review]

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 1 above)
Comment (elfenpiff) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:208]: We should remove them completely and should always use `units::duration`.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl [hunk 3]
```diff
@@ -151,19 +150,8 @@ inline bool ChunkSender<ChunkSenderDataType>::getChunkReadyForSend(const mepoo::
...
-            getMembers()->m_sequenceNumber++;
-        }
-        else
-        {
-            // if the seqence number in the chunk info is set by the user, we still increment the internal one as this
-            // might be needed by midddleware spcific evaluation (like in introspection)
-            getMembers()->m_sequenceNumber++;
-        }
+        chunk.getChunkHeader()->m_sequenceNumber = getMembers()->m_sequenceNumber;
+        getMembers()->m_sequenceNumber++;
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl:154]: Why post-increment when you are not using the previous value? Could you please use the pre increment operator.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl (hunk 2 above)
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl:45]: yes, that's part of the next PR

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl (hunk 3 above)
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl:154]: good question. legacy code 😁

Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl:154]: if there are no other issue, are you okay to fix this with the next PR?

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 1 above)
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:208]: so, no chrono based timepoints?

Comment (dkroenke) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:208]: Is the issue in cxx::Duration with the rounding error resolved? For example when you use 1 ns then its represented as zero in the duration.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/introspection/port_introspection.hpp [hunk 4]
```diff
@@ -83,10 +83,10 @@ class PortIntrospection
...
             NodeName_t node;
 
-            using TimePointNs = mepoo::TimePointNs;
-            using DurationNs = mepoo::DurationNs;
-            TimePointNs m_sequenceNumberTimestamp = TimePointNs(DurationNs(0));
-            mepoo::SequenceNumberType m_sequenceNumber{0};
+            using TimePointNs_t = mepoo::TimePointNs_t;
+            using DurationNs_t = mepoo::DurationNs_t;
+            TimePointNs_t m_sequenceNumberTimestamp {DurationNs_t(0)};
+            mepoo::SequenceNumber_t m_sequenceNumber{0};
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/introspection/port_introspection.hpp]: legacy code ;)

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl (hunk 3 above)
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl]: done

**Review by dkroenke — APPROVED** [review]

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/introspection/port_introspection.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/introspection/port_introspection.hpp]: oh no, I also did if for DurationNs_t which is a singned integer

**Review by budrus — DISMISSED** [review]

File: iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp [hunk 5]
```diff
@@ -54,7 +54,7 @@ struct ChunkDistributorConfig
 
 struct ChunkQueueConfig
 {
-    static constexpr uint64_t MAX_QUEUE_CAPACITY = NUM_CHUNKS_IN_POOL;
+    static constexpr uint64_t MAX_QUEUE_CAPACITY = iox::MAX_SUBSCRIBER_QUEUE_CAPACITY;
```
Comment (elBoberido) [inline iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:57]: @budrus @mossmaurice can you please confirm that this makes sense?

Comment (budrus) [reply iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:57]: I feel that this was chosen for a reason. With `NUM_CHUNKS_IN_POOL`, we avoid queue overflows (which are also checked in the test and lead to fail). I think we have timing issues with the Mac sanitize builds. If the forward and subscriber thread are not fast enough you either have a queue overflow or a mempool running out of chunks error. With this change you get the latter one but it still fails if the threads get no runtime

Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:57]: but then we exhaust the mempool, since there is
- `lastChunk` in publisher
- two queues which can hold `NUM_CUNKS_IN_POOL`

The overflow problem was solved by using a FIFO instead of a SOFI

Comment (budrus) [reply iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:57]: When the other threads are not fast enough it will break the one or the other way. You can now configure how it fails. SOFI or FIFO is also don't care as we should set the overflow flag also for SOFI and the test fails if there is on overflow

Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:57]: you're right. But this makes my PR green :D

**Review by budrus — APPROVED** [review]

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 1 above)
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:208]: Not, it's still using a `double`. @elfenpiff will receive the honour I suppose :P

File: iceoryx_posh/source/mepoo/chunk_header.cpp [hunk 6]
```diff
@@ -43,5 +43,14 @@ ChunkHeader* ChunkHeader::fromPayload(const void* const payload) noexcept
     return reinterpret_cast<ChunkHeader*>(payloadAddress - *payloadOffset);
 }
 
+uint32_t ChunkHeader::usedSizeOfChunk()
+{
+    auto usedSizeOfChunk = static_cast<uint64_t>(m_payloadOffset) + static_cast<uint64_t>(m_payloadSize);
+
+    cxx::Expects(usedSizeOfChunk <= m_chunkSize);
```
Comment (mossmaurice) [inline iceoryx_posh/source/mepoo/chunk_header.cpp:50]: :partying_face:

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp [hunk 7]
```diff
@@ -101,3 +103,29 @@ TEST_F(ChunkHeader_test, FromPayloadFunctionCalledWithNullptrReturnsNullptr)
...
+TEST_F(ChunkHeader_test, usedChunkSizeIsSizeOfChunkHeaderPlusOneWhenPayloadIsOne)
+{
+    ChunkHeader sut;
+    sut.m_chunkSize = 2 * sizeof(ChunkHeader);
+    constexpr uint32_t PAYLOAD_SIZE{1U};
+    sut.m_payloadSize = PAYLOAD_SIZE;
+    EXPECT_THAT(sut.usedSizeOfChunk(), Eq(sizeof(ChunkHeader) + PAYLOAD_SIZE));
+}
+
+TEST_F(ChunkHeader_test, usedChunkSizeTerminatesWhenPayloadSizeHasSillyValue)
```
Comment (mossmaurice) [inline iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp:123]: One could argue what exactly "Silly" is. Maybe rename the test case and name the variable `sillyPayload`?

**Review by mossmaurice — CHANGES_REQUESTED** [review]
Please try to avoid using `chrono` directly. The advantage of using `cxx::Duration` is that we have conversion operators to the commonly used `timeval` and `timespec`. I suppose the best idea is to use `chrono` inside `cxx::Duration`

File: iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp (hunk 5 above)
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:57]: yesterday two other PRs had this problem. It seems they are building now. I will revert it and check if it builds now. If not, we should try for some time with a smaller queue size

File: iceoryx_posh/test/moduletests/test_mepoo_segment.cpp [hunk 8]
```diff
@@ -184,5 +184,5 @@ TEST_F(MePooSegment_test, ADD_TEST_WITH_ADDITIONAL_USER(GetMemoryManager))
     auto config = sut.getMemoryManager().getMemPoolInfo(0);
     ASSERT_THAT(config.m_numChunks, Eq(100u));
     auto chunk = sut.getMemoryManager().getChunk(128);
-    EXPECT_THAT(chunk.getChunkHeader()->m_info.m_payloadSize, Eq(128u));
+    EXPECT_THAT(chunk.getChunkHeader()->m_payloadSize, Eq(128u));
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_mepoo_segment.cpp]: three lowercase u replaced, 489 remaining

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp (hunk 7 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp]: done

**Comment by elBoberido** [comment]
@mossmaurice the usage of `chrono` is not new but legacy code. It was used for the timestamp in the `ChunkInfo` and is still used for the timestamp of the last heartbeat of the application. We don't have something to replace that. If the decision is to get rid of chrono, I would create an issue for that.

**Comment by mossmaurice** [comment]
> @mossmaurice the usage of `chrono` is not new but legacy code. It was used for the timestamp in the `ChunkInfo` and is still used for the timestamp of the last heartbeat of the application. We don't have something to replace that. If the decision is to get rid of chrono, I would create an issue for that.

Fine by me. @budrus What about tackling this topic in v1.x? Or is `chrono` and `cxx::Duration` used in a public API?

File: iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp (hunk 5 above)
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp]: okay, I will revert my revert. The original test now fails again

**Comment by elBoberido** [comment]
@mossmaurice chrono is not user facing. I'm not sure about `cxx::Duration`, but I think we use it everywhere a timeout is specified

File: iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp (hunk 5 above)
Comment (budrus) [reply iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:57]: You are hunting a ghost. Both versions will fail. It's just gambling. We have to change the test or somehow disable it in the sanitize build. I'm pretty sure that it is a timing issue

Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:57]: you are totally right, but for some reasons I had 4 red builds without this change and 2 greed builds with this change ... is Vegas still open?

**Review by budrus — APPROVED** [review]

**Comment by budrus** [comment]
> 
> 
> > @mossmaurice the usage of `chrono` is not new but legacy code. It was used for the timestamp in the `ChunkInfo` and is still used for the timestamp of the last heartbeat of the application. We don't have something to replace that. If the decision is to get rid of chrono, I would create an issue for that.
> 
> Fine by me. @budrus What about tackling this topic in v1.x? Or is `chrono` and `cxx::Duration` used in a public API?

Does it makes sense to address this together with #337?

**Comment by elBoberido** [comment]
> > > @mossmaurice the usage of `chrono` is not new but legacy code. It was used for the timestamp in the `ChunkInfo` and is still used for the timestamp of the last heartbeat of the application. We don't have something to replace that. If the decision is to get rid of chrono, I would create an issue for that.
> > 
> > 
> > Fine by me. @budrus What about tackling this topic in v1.x? Or is `chrono` and `cxx::Duration` used in a public API?
> 
> Does it makes sense to address this together with #337?

yes, I added it there

**Review by mossmaurice — APPROVED** [review]
